### PR TITLE
lightning/external: add statistic for multiple data files

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #45719

Problem Summary:

### What is changed and how it works?

Because all statistic files are too large to be loaded into one node, we introduce second-level statistic file that describes the property every 500 files.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/external/writer.go [hunk 1]
```diff
@@ -173,7 +176,47 @@ func (b *WriterBuilder) Build(
...
+	return ret
+}
+
+// MultipleFilesStat is the statistic information of multiple files (currently
+// every 500 files). It is used to estimate the data overlapping, and per-file
+// statistic information maybe too big to loaded into memory.
+type MultipleFilesStat struct {
+	MinKey            tidbkv.Key
+	MaxKey            tidbkv.Key
+	Filenames         [][2]string // [dataFile, statFile]
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/external/writer.go:193]: Is one name enough?

File: br/pkg/lightning/backend/external/util.go [hunk 2]
```diff
@@ -186,3 +187,50 @@ func MockExternalEngine(
 	}
 	return GetAllFileNames(ctx, storage, "/mock-test")
 }
+
+// EndpointTp is the type of Endpoint.Key.
+type EndpointTp int
+
+const (
+	// ExclusiveEnd represents "..., Endpoint.Key)".
+	ExclusiveEnd EndpointTp = iota
```
Comment (wjhuang2016) [inline br/pkg/lightning/backend/external/util.go:196]: I guess we don't need it?

File: br/pkg/lightning/backend/external/writer.go (hunk 1 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/writer.go:193]: I have no strong opinion. If in future we won't have multiple data files share the same state file, I think we can remove it and provide a helper function to convert filenames.
And if you think the serialized metadata is too large, then it's a must to only keep filenames of data file. I'm not sure about it.

File: br/pkg/lightning/backend/external/util.go (hunk 2 above)
Comment (lance6716) [reply br/pkg/lightning/backend/external/util.go:196]: Yes currently we don't need it. But I'm not sure there will be usage in future like "checking the overlap of RangeSplitter" or something.

File: br/pkg/lightning/backend/external/util.go [hunk 3]
```diff
@@ -186,3 +187,50 @@ func MockExternalEngine(
...
+	Weight uint64 // all EndpointTp use positive weight
+}
+
+// GetMaxOverlapping returns the maximum overlapping weight treating given
+// `points` as endpoints of intervals. `points` are not required to be sorted,
+// and will be sorted in-place in this function.
+func GetMaxOverlapping(points []Endpoint) int {
+	slices.SortFunc(points, func(i, j Endpoint) int {
+		if bytes.Compare(i.Key, j.Key) != 0 {
+			return bytes.Compare(i.Key, j.Key)
```
Comment (tangenta) [inline br/pkg/lightning/backend/external/util.go:216]: Can we avoid comparing twice?

Comment (lance6716) [reply br/pkg/lightning/backend/external/util.go]: https://github.com/pingcap/tidb/pull/46614/commits/3377a054f4fa40ed1a4404255b7cd0b271b1274b

**Review by wjhuang2016 — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**